## Introduction

The PedPheWAS R package is designed to perform phenome-wide association
studies (PheWAS) using pediatric phecodes (Ped-Phecodes).

This package was built based on the original PheWAS R package available
at: <https://github.com/PheWAS/PheWAS>

## Installation

To install the latest development version of the PedPheWAS R package,
use the following import statements, or run the packageSetup function:

    # install and load required packages 
    install.packages(c("devtools","vroom","tidyverse","dplyr"))

    ## Error in install.packages : Updating loaded packages

    library(devtools)
    library(vroom)
    library(tidyverse)
    library(dplyr)
    library(parallel)

    devtools::install_github("monikagrabowska/PediatricPheCodes")

    ## Using GitHub PAT from the git credential store.

    ## Skipping install of 'PediatricPheCodes' from a github remote, the SHA1 (5a9bef15) has not changed since last install.
    ##   Use `force = TRUE` to force installation

    library(PediatricPheCodes)

## Tutorial

This tutorial demonstrates how to use the PedPheWAS R package to perform
a PheWAS analysis using simulated data.

### 1) Prepare and load input data

To perform a PheWAS, the user must provide three data frames: 1) a data
frame containing patient ICD-9-CM and ICD-10-CM codes and code counts
(or code index dates), 2) a data frame with patient sex (used to
determine case/control/exclude status for sex-specific phecodes), and 3)
a data frame containing the patient genotypes.

An example of each of these data frames is provided below, showing the
required column formatting.

If you start by running the packageSetup function, these dataframes will
be previously defined

    example_data_icd <- vroom::vroom("./data/person_icd.csv",
    .name = janitor::make_clean_names, delim = ",",
    col_types = c(id = "i", vocabulary_id = "c", code = "c", count = "i"))

    example_data_sex <- vroom::vroom("./data/test_sex.csv",
    .name = janitor::make_clean_names, delim = ",",
    col_types = c(id = "i", sex = "c"))

    example_data_genotypes <- vroom::vroom("./data/test_genotypes.csv",
    .name = janitor::make_clean_names, delim = ",",
    col_types = c(id = "i", rs_example = "i"))

### 2) Mapping ICD-9-CM and ICD-10-CM codes to pediatric phecodes

The createPedPhenotypes() function maps the patient ICD-9-CM and
ICD-10-CM codes to pediatric phecodes.

    ped_phenotypes <- createPedPhenotypes(example_data_icd, example_data_sex)

    ## Mapping codes to phecodes...

    ## Aggregating codes...

    ## Mapping exclusions...

    ## Coalescing exclusions and min.code.count as applicable...

    ## Reshaping data...

The function returns a data frame of boolean phenotypes, in which the
first column contains patient IDs and the remaining columns indicate
case/control/exclude status (i.e., TRUE/FALSE/NA) for each phecode.

A subset of the example output is shown below:

    #*insert plot here*

### 3) Preparing data for PheWAS

Join the patient genotype data to the boolean phenotypes generated by
the createPedPhenotypes() function to create the data frame needed to
perform PheWAS. If planning to adjust for covariates (e.g., age,
gender), these should also be joined to the data frame.

    df_full <- dplyr::inner_join(example_data_genotypes, ped_phenotypes, by = "id")
    df_full

    ## # A tibble: 5,000 × 2,009
    ##       id rs_example `008` `008.5` `008.51` `008.511` `008.512` `008.513` `008.514` `008.515` `008.516`
    ##    <int>      <int> <lgl> <lgl>   <lgl>    <lgl>     <lgl>     <lgl>     <lgl>     <lgl>     <lgl>    
    ##  1     1          1 FALSE FALSE   FALSE    FALSE     FALSE     FALSE     FALSE     FALSE     FALSE    
    ##  2     2          0 FALSE FALSE   FALSE    FALSE     FALSE     FALSE     FALSE     FALSE     FALSE    
    ##  3     3          1 FALSE FALSE   FALSE    FALSE     FALSE     FALSE     FALSE     FALSE     FALSE    
    ##  4     4          1 FALSE FALSE   FALSE    FALSE     FALSE     FALSE     FALSE     FALSE     FALSE    
    ##  5     5          1 FALSE FALSE   FALSE    FALSE     FALSE     FALSE     FALSE     FALSE     FALSE    
    ##  6     6          1 FALSE FALSE   FALSE    FALSE     FALSE     FALSE     FALSE     FALSE     FALSE    
    ##  7     7          1 FALSE FALSE   FALSE    FALSE     FALSE     FALSE     FALSE     FALSE     FALSE    
    ##  8     8          1 FALSE FALSE   FALSE    FALSE     FALSE     FALSE     FALSE     FALSE     FALSE    
    ##  9     9          2 FALSE FALSE   FALSE    FALSE     FALSE     FALSE     FALSE     FALSE     FALSE    
    ## 10    10          1 FALSE FALSE   FALSE    FALSE     FALSE     FALSE     FALSE     FALSE     FALSE    
    ## # ℹ 4,990 more rows
    ## # ℹ 1,998 more variables: `008.52` <lgl>, `008.6` <lgl>, `008.7` <lgl>, `010` <lgl>, `031` <lgl>,
    ## #   `031.1` <lgl>, `038` <lgl>, `038.1` <lgl>, `038.2` <lgl>, `038.3` <lgl>, `041` <lgl>,
    ## #   `041.1` <lgl>, `041.11` <lgl>, `041.12` <lgl>, `041.2` <lgl>, `041.21` <lgl>, `041.4` <lgl>,
    ## #   `041.8` <lgl>, `041.9` <lgl>, `053` <lgl>, `054` <lgl>, `070` <lgl>, `070.1` <lgl>, `070.2` <lgl>,
    ## #   `070.3` <lgl>, `070.4` <lgl>, `070.9` <lgl>, `071` <lgl>, `071.1` <lgl>, `078` <lgl>, `079` <lgl>,
    ## #   `079.1` <lgl>, `079.2` <lgl>, `079.3` <lgl>, `079.4` <lgl>, `079.5` <lgl>, `079.6` <lgl>, …

### 4) Performing PheWAS

The phewas\_ext() function performs the PheWAS using the data frame with
phenotype, genotype, and any covariate data.

    results <- phewas_ext(phenotypes=names(ped_phenotypes)[-1], genotypes=colnames(example_data_genotypes)[2], df_full)

    ## Finding associations...

    ## Compiling results...

    ## Cleaning up...

A subset of the example output is shown below:

    results

    ##                 phenotype        snp covariates        beta        SE        OR          p     type
    ## rs_example...1        008 rs_example       <NA> -0.18988021 0.1740896 0.8270582 0.27540330 logistic
    ## rs_example...2      008.5 rs_example       <NA> -0.03040417 0.2388391 0.9700534 0.89870310 logistic
    ## rs_example...3     008.51 rs_example       <NA> -0.40554493 0.4577433 0.6666135 0.37563589 logistic
    ## rs_example...4    008.511 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...5    008.512 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...6    008.513 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...7    008.514 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...8    008.515 rs_example       <NA>          NA        NA        NA         NA     <NA>
    ## rs_example...9    008.516 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...10    008.52 rs_example       <NA>  0.01432655 0.3891918 1.0144297 0.97063568 logistic
    ## rs_example...11     008.6 rs_example       <NA> -0.19908424 0.4522326 0.8194809 0.65977389 logistic
    ## rs_example...12     008.7 rs_example       <NA> -0.43499146 0.4223748 0.6472702 0.30307062 logistic
    ## rs_example...13       010 rs_example       <NA>  0.09660178 0.4741738 1.1014217 0.83856721 logistic
    ## rs_example...14       031 rs_example       <NA> -0.08001760 0.3212643 0.9231001 0.80330588 logistic
    ## rs_example...15     031.1 rs_example       <NA> -0.60985629 0.4818710 0.5434290 0.20565605 logistic
    ## rs_example...16       038 rs_example       <NA> -0.04051374 0.1819908 0.9602960 0.82383581 logistic
    ## rs_example...17     038.1 rs_example       <NA>  0.08819426 0.4085228 1.0922003 0.82907680 logistic
    ## rs_example...18     038.2 rs_example       <NA> -0.50104244 0.3292940 0.6058987 0.12811795 logistic
    ## rs_example...19     038.3 rs_example       <NA> -0.36212505 0.3520944 0.6961953 0.30372010 logistic
    ## rs_example...20       041 rs_example       <NA>  0.05510935 0.1485142 1.0566562 0.71058442 logistic
    ## rs_example...21     041.1 rs_example       <NA> -0.10378875 0.2670531 0.9014157 0.69753900 logistic
    ## rs_example...22    041.11 rs_example       <NA>  0.06440546 0.4639072 1.0665247 0.88958240 logistic
    ## rs_example...23    041.12 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...24     041.2 rs_example       <NA> -0.10226131 0.3157280 0.9027936 0.74602090 logistic
    ## rs_example...25    041.21 rs_example       <NA> -0.03949209 0.4297676 0.9612776 0.92678407 logistic
    ## rs_example...26     041.4 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...27     041.8 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...28     041.9 rs_example       <NA>  0.68987671 0.4069414 1.9934697 0.09002366 logistic
    ## rs_example...29       053 rs_example       <NA>  0.39896709 0.2672624 1.4902846 0.13549176 logistic
    ## rs_example...30       054 rs_example       <NA> -0.39278000 0.4555755 0.6751773 0.38859819 logistic
    ## rs_example...31       070 rs_example       <NA> -0.13821215 0.1655182 0.8709139 0.40370253 logistic
    ## rs_example...32     070.1 rs_example       <NA>  0.14111167 0.4191639 1.1515532 0.73638050 logistic
    ## rs_example...33     070.2 rs_example       <NA> -0.26036527 0.4100012 0.7707700 0.52540529 logistic
    ## rs_example...34     070.3 rs_example       <NA> -0.60057506 0.4011252 0.5484961 0.13433449 logistic
    ## rs_example...35     070.4 rs_example       <NA> -0.42284223 0.4208470 0.6551820 0.31502159 logistic
    ## rs_example...36     070.9 rs_example       <NA>  0.41954180 0.3658270 1.5212644 0.25145143 logistic
    ## rs_example...37       071 rs_example       <NA> -0.07674429 0.3564276 0.9261266 0.82952154 logistic
    ## rs_example...38     071.1 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...39       078 rs_example       <NA>  0.04391119 0.3935432 1.0448896 0.91115716 logistic
    ## rs_example...40       079 rs_example       <NA>  0.13601244 0.1892345 1.1456961 0.47229445 logistic
    ## rs_example...41     079.1 rs_example       <NA>  0.01034643 0.4428769 1.0104001 0.98136163 logistic
    ## rs_example...42     079.2 rs_example       <NA> -0.03130735 0.3720945 0.9691777 0.93294657 logistic
    ## rs_example...43     079.3 rs_example       <NA>          NA        NA        NA         NA     <NA>
    ## rs_example...44     079.4 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...45     079.5 rs_example       <NA>          NA        NA        NA         NA     <NA>
    ## rs_example...46     079.6 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...47     079.7 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...48     079.8 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...49     079.9 rs_example       <NA>  0.40535717 0.3926075 1.4998381 0.30184990 logistic
    ## rs_example...50     079.A rs_example       <NA>          NA        NA        NA         NA     <NA>
    ## rs_example...51     079.B rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...52     079.C rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...53       080 rs_example       <NA>          NA        NA        NA         NA logistic
    ## rs_example...54       081 rs_example       <NA>  0.26846158 0.2331878 1.3079507 0.24962222 logistic
    ## rs_example...55     081.1 rs_example       <NA>  0.28889057 0.2355887 1.3349456 0.22010479 logistic
    ## rs_example...56    081.11 rs_example       <NA>  0.20923897 0.4340433 1.2327396 0.62975670 logistic
    ## rs_example...57    081.12 rs_example       <NA>  0.20959116 0.4343765 1.2331738 0.62944345 logistic
    ## rs_example...58       090 rs_example       <NA>  0.28985511 0.2219525 1.3362339 0.19157527 logistic
    ##                 n_total n_cases n_controls HWE_p allele_freq n_no_snp                  formula
    ## rs_example...1     4998     158       4840     1   0.5644258        0     `008` ~ `rs_example`
    ## rs_example...2     4921      81       4840     1   0.5650274        0   `008.5` ~ `rs_example`
    ## rs_example...3     4863      23       4840     1   0.5648776        0  `008.51` ~ `rs_example`
    ## rs_example...4     4843       3       4840     1   0.5651456        0 `008.511` ~ `rs_example`
    ## rs_example...5     4841       1       4840     1   0.5650692        0 `008.512` ~ `rs_example`
    ## rs_example...6     4848       8       4840     1   0.5649752        0 `008.513` ~ `rs_example`
    ## rs_example...7     4843       3       4840     1   0.5650423        0 `008.514` ~ `rs_example`
    ## rs_example...8     4840      NA         NA    NA          NA        0                     <NA>
    ## rs_example...9     4844       4       4840     1   0.5650289        0 `008.516` ~ `rs_example`
    ## rs_example...10    4870      30       4840     1   0.5650924        0  `008.52` ~ `rs_example`
    ## rs_example...11    4863      23       4840     1   0.5649805        0   `008.6` ~ `rs_example`
    ## rs_example...12    4867      27       4840     1   0.5648243        0   `008.7` ~ `rs_example`
    ## rs_example...13    4599      20       4579     1   0.5643618        0     `010` ~ `rs_example`
    ## rs_example...14    4624      45       4579     1   0.5642301        0     `031` ~ `rs_example`
    ## rs_example...15    4600      21       4579     1   0.5640217        0   `031.1` ~ `rs_example`
    ## rs_example...16    4721     142       4579     1   0.5641813        0     `038` ~ `rs_example`
    ## rs_example...17    4606      27       4579     1   0.5643726        0   `038.1` ~ `rs_example`
    ## rs_example...18    4624      45       4579     1   0.5637976        0   `038.2` ~ `rs_example`
    ## rs_example...19    4618      39       4579     1   0.5639887        0   `038.3` ~ `rs_example`
    ## rs_example...20    4792     213       4579     1   0.5645868        0     `041` ~ `rs_example`
    ## rs_example...21    4645      66       4579     1   0.5641550        0   `041.1` ~ `rs_example`
    ## rs_example...22    4600      21       4579     1   0.5643478        0  `041.11` ~ `rs_example`
    ## rs_example...23    4598      19       4579     1   0.5641583        0  `041.12` ~ `rs_example`
    ## rs_example...24    4626      47       4579     1   0.5642023        0   `041.2` ~ `rs_example`
    ## rs_example...25    4604      25       4579     1   0.5642919        0  `041.21` ~ `rs_example`
    ## rs_example...26    4597      18       4579     1   0.5646074        0   `041.4` ~ `rs_example`
    ## rs_example...27    4591      12       4579     1   0.5641472        0   `041.8` ~ `rs_example`
    ## rs_example...28    4603      24       4579     1   0.5647404        0   `041.9` ~ `rs_example`
    ## rs_example...29    4609      59       4550     1   0.5646561        0     `053` ~ `rs_example`
    ## rs_example...30    4573      23       4550     1   0.5638531        0     `054` ~ `rs_example`
    ## rs_example...31    4724     174       4550     1   0.5635055        0     `070` ~ `rs_example`
    ## rs_example...32    4575      25       4550     1   0.5641530        0   `070.1` ~ `rs_example`
    ## rs_example...33    4578      28       4550     1   0.5638925        0   `070.2` ~ `rs_example`
    ## rs_example...34    4580      30       4550     1   0.5636463        0   `070.3` ~ `rs_example`
    ## rs_example...35    4577      27       4550     1   0.5637972        0   `070.4` ~ `rs_example`
    ## rs_example...36    4581      31       4550     1   0.5643964        0   `070.9` ~ `rs_example`
    ## rs_example...37    4586      36       4550     1   0.5639991        0     `071` ~ `rs_example`
    ## rs_example...38    4559       9       4550     1   0.5640491        0   `071.1` ~ `rs_example`
    ## rs_example...39    4579      29       4550     1   0.5640970        0     `078` ~ `rs_example`
    ## rs_example...40    4676     126       4550     1   0.5644782        0     `079` ~ `rs_example`
    ## rs_example...41    4573      23       4550     1   0.5640717        0   `079.1` ~ `rs_example`
    ## rs_example...42    4583      33       4550     1   0.5640410        0   `079.2` ~ `rs_example`
    ## rs_example...43    4869      NA         NA    NA          NA        0                     <NA>
    ## rs_example...44    4879      10       4869     1   0.5637426        0   `079.4` ~ `rs_example`
    ## rs_example...45    4869      NA         NA    NA          NA        0                     <NA>
    ## rs_example...46    4871       2       4869     1   0.5637446        0   `079.6` ~ `rs_example`
    ## rs_example...47    4871       2       4869     1   0.5637446        0   `079.7` ~ `rs_example`
    ## rs_example...48    4870       1       4869     1   0.5637577        0   `079.8` ~ `rs_example`
    ## rs_example...49    4577      27       4550     1   0.5643435        0   `079.9` ~ `rs_example`
    ## rs_example...50    4869      NA         NA    NA          NA        0                     <NA>
    ## rs_example...51    4872       3       4869     1   0.5637315        0   `079.B` ~ `rs_example`
    ## rs_example...52    4870       1       4869     1   0.5637577        0   `079.C` ~ `rs_example`
    ## rs_example...53    4916      18       4898     1   0.5637714        0     `080` ~ `rs_example`
    ## rs_example...54    4978      80       4898     1   0.5636802        0     `081` ~ `rs_example`
    ## rs_example...55    4976      78       4898     1   0.5637058        0   `081.1` ~ `rs_example`
    ## rs_example...56    4921      23       4898     1   0.5633001        0  `081.11` ~ `rs_example`
    ## rs_example...57    4921      23       4898     1   0.5633001        0  `081.12` ~ `rs_example`
    ## rs_example...58    4994      88       4906     1   0.5640769        0     `090` ~ `rs_example`
    ##                         expanded_formula
    ## rs_example...1  (Intercept) + rs_example
    ## rs_example...2  (Intercept) + rs_example
    ## rs_example...3  (Intercept) + rs_example
    ## rs_example...4                      <NA>
    ## rs_example...5                      <NA>
    ## rs_example...6                      <NA>
    ## rs_example...7                      <NA>
    ## rs_example...8                      <NA>
    ## rs_example...9                      <NA>
    ## rs_example...10 (Intercept) + rs_example
    ## rs_example...11 (Intercept) + rs_example
    ## rs_example...12 (Intercept) + rs_example
    ## rs_example...13 (Intercept) + rs_example
    ## rs_example...14 (Intercept) + rs_example
    ## rs_example...15 (Intercept) + rs_example
    ## rs_example...16 (Intercept) + rs_example
    ## rs_example...17 (Intercept) + rs_example
    ## rs_example...18 (Intercept) + rs_example
    ## rs_example...19 (Intercept) + rs_example
    ## rs_example...20 (Intercept) + rs_example
    ## rs_example...21 (Intercept) + rs_example
    ## rs_example...22 (Intercept) + rs_example
    ## rs_example...23                     <NA>
    ## rs_example...24 (Intercept) + rs_example
    ## rs_example...25 (Intercept) + rs_example
    ## rs_example...26                     <NA>
    ## rs_example...27                     <NA>
    ## rs_example...28 (Intercept) + rs_example
    ## rs_example...29 (Intercept) + rs_example
    ## rs_example...30 (Intercept) + rs_example
    ## rs_example...31 (Intercept) + rs_example
    ## rs_example...32 (Intercept) + rs_example
    ## rs_example...33 (Intercept) + rs_example
    ## rs_example...34 (Intercept) + rs_example
    ## rs_example...35 (Intercept) + rs_example
    ## rs_example...36 (Intercept) + rs_example
    ## rs_example...37 (Intercept) + rs_example
    ## rs_example...38                     <NA>
    ## rs_example...39 (Intercept) + rs_example
    ## rs_example...40 (Intercept) + rs_example
    ## rs_example...41 (Intercept) + rs_example
    ## rs_example...42 (Intercept) + rs_example
    ## rs_example...43                     <NA>
    ## rs_example...44                     <NA>
    ## rs_example...45                     <NA>
    ## rs_example...46                     <NA>
    ## rs_example...47                     <NA>
    ## rs_example...48                     <NA>
    ## rs_example...49 (Intercept) + rs_example
    ## rs_example...50                     <NA>
    ## rs_example...51                     <NA>
    ## rs_example...52                     <NA>
    ## rs_example...53                     <NA>
    ## rs_example...54 (Intercept) + rs_example
    ## rs_example...55 (Intercept) + rs_example
    ## rs_example...56 (Intercept) + rs_example
    ## rs_example...57 (Intercept) + rs_example
    ## rs_example...58 (Intercept) + rs_example
    ##                                                                                                                        note
    ## rs_example...1                                                                                                             
    ## rs_example...2                                                                                                             
    ## rs_example...3                                                                                                             
    ## rs_example...4                                                                             [Error: < 20  cases or controls]
    ## rs_example...5                                                                             [Error: < 20  cases or controls]
    ## rs_example...6                                                                             [Error: < 20  cases or controls]
    ## rs_example...7                                                                             [Error: < 20  cases or controls]
    ## rs_example...8   [Note: Column(s) dropped due to lack of variability:  008.515 ] [Error: non-varying phenotype or genotype]
    ## rs_example...9                                                                             [Error: < 20  cases or controls]
    ## rs_example...10                                                                                                            
    ## rs_example...11                                                                                                            
    ## rs_example...12                                                                                                            
    ## rs_example...13                                                                                                            
    ## rs_example...14                                                                                                            
    ## rs_example...15                                                                                                            
    ## rs_example...16                                                                                                            
    ## rs_example...17                                                                                                            
    ## rs_example...18                                                                                                            
    ## rs_example...19                                                                                                            
    ## rs_example...20                                                                                                            
    ## rs_example...21                                                                                                            
    ## rs_example...22                                                                                                            
    ## rs_example...23                                                                            [Error: < 20  cases or controls]
    ## rs_example...24                                                                                                            
    ## rs_example...25                                                                                                            
    ## rs_example...26                                                                            [Error: < 20  cases or controls]
    ## rs_example...27                                                                            [Error: < 20  cases or controls]
    ## rs_example...28                                                                                                            
    ## rs_example...29                                                                                                            
    ## rs_example...30                                                                                                            
    ## rs_example...31                                                                                                            
    ## rs_example...32                                                                                                            
    ## rs_example...33                                                                                                            
    ## rs_example...34                                                                                                            
    ## rs_example...35                                                                                                            
    ## rs_example...36                                                                                                            
    ## rs_example...37                                                                                                            
    ## rs_example...38                                                                            [Error: < 20  cases or controls]
    ## rs_example...39                                                                                                            
    ## rs_example...40                                                                                                            
    ## rs_example...41                                                                                                            
    ## rs_example...42                                                                                                            
    ## rs_example...43    [Note: Column(s) dropped due to lack of variability:  079.3 ] [Error: non-varying phenotype or genotype]
    ## rs_example...44                                                                            [Error: < 20  cases or controls]
    ## rs_example...45    [Note: Column(s) dropped due to lack of variability:  079.5 ] [Error: non-varying phenotype or genotype]
    ## rs_example...46                                                                            [Error: < 20  cases or controls]
    ## rs_example...47                                                                            [Error: < 20  cases or controls]
    ## rs_example...48                                                                            [Error: < 20  cases or controls]
    ## rs_example...49                                                                                                            
    ## rs_example...50    [Note: Column(s) dropped due to lack of variability:  079.A ] [Error: non-varying phenotype or genotype]
    ## rs_example...51                                                                            [Error: < 20  cases or controls]
    ## rs_example...52                                                                            [Error: < 20  cases or controls]
    ## rs_example...53                                                                            [Error: < 20  cases or controls]
    ## rs_example...54                                                                                                            
    ## rs_example...55                                                                                                            
    ## rs_example...56                                                                                                            
    ## rs_example...57                                                                                                            
    ## rs_example...58                                                                                                            
    ##  [ reached 'max' / getOption("max.print") -- omitted 1949 rows ]

### 5) Plotting PheWAS results

The results of the PheWAS can be visualized in a Manhattan plot using
the plotManhattan() function. The Manhattan plot shows the negative log
p-value of the associations, with phecodes arranged on the x-axis by
disease category. Annotations can be added to the plot to highlight
notable associations.

1.  Manhattan plot with annotations specified using annotate.level:

<!-- -->

    test_plot1 <- plotManhattan(results, annotate.phenotype = T, annotate.level = 0.005, y.axis.interval = 1)

    test_plot1

![](README_files/figure-markdown_strict/step5a-1.png)

1.  Manhattan plot with annotations specified using annotate.list:

<!-- -->

    test_plot2 <- plotManhattan(results, annotate.phenotype = T, annotate.list = c("622"), y.axis.interval = 1)

    test_plot2

![](README_files/figure-markdown_strict/step5b-1.png)

1.  Manhattan plot without annotations:

<!-- -->

    test_plot3 <- plotManhattan(results, annotate.phenotype = F, y.axis.interval = 1)

    test_plot3

![](README_files/figure-markdown_strict/step5c-1.png)

## Citation

**insert citations here**
